client/lease.go

```diff
@@ -34,82 +84,141 @@ func newLeaseClient(
+func (c *leaseClient) Grant(ttl int64, opts ...LeaseOption) (*LeaseGrantResponse, error) {
+	request := toGrantReq(ttl, opts...)
 	if request.ID == 0 {
 		request.ID = c.idGen.next()
 	}
-	requestWithToken := xlineapi.RequestWithToken{
+	requestWithToken := pb.RequestWithToken{
 		Token: &c.token,
-		RequestWrapper: &xlineapi.RequestWithToken_LeaseGrantRequest{
+		RequestWrapper: &pb.RequestWithToken_LeaseGrantRequest{
 			LeaseGrantRequest: request,
 		},
 	}
 	proposeId := c.generateProposeId()
-	cmd := xlineapi.Command{Request: &requestWithToken, ProposeId: proposeId}
+	cmd := pb.Command{Request: &requestWithToken, ProposeId: proposeId}
 
 	res, err := c.curpClient.propose(&cmd, true)
 	if err != nil {
 		return nil, err
 	}
-	return res.CommandResp.GetLeaseGrantResponse(), err
+	return (*LeaseGrantResponse)(res.CommandResp.GetLeaseGrantResponse()), err
 }
 
 // Revokes a lease. All keys attached to the lease will expire and be deleted.
-func (c *leaseClient) Revoke(request *xlineapi.LeaseRevokeRequest) (*xlineapi.LeaseRevokeResponse, error) {
-	requestWithToken := xlineapi.RequestWithToken{
+func (c *leaseClient) Revoke(id int64) (*LeaseRevokeResponse, error) {
+	requestWithToken := pb.RequestWithToken{
 		Token: &c.token,
-		RequestWrapper: &xlineapi.RequestWithToken_LeaseRevokeRequest{
-			LeaseRevokeRequest: request,
+		RequestWrapper: &pb.RequestWithToken_LeaseRevokeRequest{
+			LeaseRevokeRequest: &pb.LeaseRevokeRequest{ID: id},
 		},
 	}
 	proposeId := c.generateProposeId()
-	cmd := xlineapi.Command{Request: &requestWithToken, ProposeId: proposeId}
+	cmd := pb.Command{Request: &requestWithToken, ProposeId: proposeId}
 
 	res, err := c.curpClient.propose(&cmd, true)
 	if err != nil {
 		return nil, err
 	}
-	return res.CommandResp.GetLeaseRevokeResponse(), err
+	return (*LeaseRevokeResponse)(res.CommandResp.GetLeaseRevokeResponse()), err
 }
 
 // Keeps the lease alive by streaming keep alive requests from the client
 // to the server and streaming keep alive responses from the server to the client.
-func (c *leaseClient) KeepAlive(request *xlineapi.LeaseKeepAliveRequest) (xlineapi.Lease_LeaseKeepAliveClient, error) {
-	keeper, err := c.leaseClient.LeaseKeepAlive(context.Background())
-	return keeper, err
+func (c *leaseClient) KeepAlive(ctx context.Context, id int64) (<-chan *LeaseKeepAliveResponse, error) {
+	kach := make(chan *LeaseKeepAliveResponse)
+
+	stream, err := c.leaseClient.LeaseKeepAlive(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				err = stream.Send(&pb.LeaseKeepAliveRequest{ID: id})
+				if err != nil {
+					c.logger.Error("keep alive fail", zap.Error(err))
+				}
+				res, err := stream.Recv()
+				if err != nil {
+					c.logger.Error("keep alive fail", zap.Error(err))
+				}
+				kach <- (*LeaseKeepAliveResponse)(res)
+			}
+			time.Sleep(1 * time.Millisecond)
```

Comment:
Could we make it configurable or a constant?

Reply:
I changed it to a constant.